File: tools/qwen36_node_reports.py
#!/usr/bin/env python3
"""Receive and summarize Qwen3.6 node packet report bundles."""
from __future__ import annotations

import argparse
import json
import os
import shutil
import subprocess
import sys
import zipfile
from pathlib import Path, PurePosixPath
from typing import Any


ROOT = Path(__file__).resolve().parents[1]
SCHEMA = "fak.qwen36-node-reports.v1"
# The operator superrepo nests the module under <root>/fak/; the standalone
# public checkout has experiments/ at the module root. Prefer whichever exists
# so the default out-dir lands in the real tree (not a phantom fak/experiments).
_BASE = ROOT / "fak" if (ROOT / "fak" / "experiments").is_dir() else ROOT
DEFAULT_INBOX = ROOT / "tools" / "_registry" / "qwen36-report-inbox"
DEFAULT_OUT_DIR = _BASE / "experiments" / "qwen36" / "node-reports"


def find_tailscale() -> str:
    exe = shutil.which("tailscale")
    if exe:
        return exe
    if os.name == "nt":
        candidate = r"C:\Program Files\Tailscale\tailscale.exe"
        if os.path.exists(candidate):
            return candidate
    return ""


def receive_taildrop(inbox: Path, wait: bool) -> dict[str, Any]:
    inbox.mkdir(parents=True, exist_ok=True)
    exe = find_tailscale()
    if not exe:
        return {"ran": False, "reason": "tailscale CLI not found", "inbox": str(inbox)}
    cmd = [exe, "file", "get", "--conflict=rename"]
    if wait:
        cmd.append("--wait")
    cmd.append(str(inbox))
    proc = subprocess.run(cmd, text=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=False)
    return {
        "ran": True,
        "command": cmd,
        "exit_code": proc.returncode,
        "stdout": proc.stdout[-2000:],
        "stderr": proc.stderr[-2000:],
        "inbox": str(inbox),
    }


def archive_candidates(inbox: Path) -> list[Path]:
    patterns = ("qwen36-node-reports-*.zip", "*qwen36-node-reports-*.zip")
    paths: list[Path] = []
    for pattern in patterns:
        paths.extend(inbox.glob(pattern))
    return sorted(set(paths), key=lambda p: p.stat().st_mtime, reverse=True)


def safe_member_name(name: str) -> PurePosixPath:
    normalized = name.replace("\\", "/")
    path = PurePosixPath(normalized)
    if path.is_absolute() or ".." in path.parts or not path.parts or any(":" in part for part in path.parts):
        raise ValueError(f"unsafe zip member path: {name}")
    return path


def validate_zip(archive: Path) -> list[zipfile.ZipInfo]:
    members: list[zipfile.ZipInfo] = []
    with zipfile.ZipFile(archive) as zf:
        for info in zf.infolist():
            safe_member_name(info.filename)
            if info.is_dir():
                continue
            members.append(info)
    if not members:
        raise ValueError(f"archive contains no files: {archive}")
    return members


def read_text_auto(path: Path) -> str:
    raw = path.read_bytes()
    if raw.startswith(b"\xff\xfe") or raw.startswith(b"\xfe\xff"):
        return raw.decode("utf-16")
    if raw.startswith(b"\xef\xbb\xbf"):
        return raw.decode("utf-8-sig")
    try:
        return raw.decode("utf-8")
    except UnicodeDecodeError:
        return raw.decode("utf-16")


def extract_archive(archive: Path, out_dir: Path, replace: bool) -> Path:
    archive = archive.resolve()
    dest = out_dir / archive.stem
    if dest.exists() and not replace:
        raise FileExistsError(f"{dest} already exists; rerun with --replace")
    if dest.exists():
        shutil.rmtree(dest)
    dest.mkdir(parents=True, exist_ok=True)
    with zipfile.ZipFile(archive) as zf:
        members = validate_zip(archive)
        for info in members:
            name = safe_member_name(info.filename)
            target = (dest / Path(*name.parts)).resolve()
            if not str(target).startswith(str(dest.resolve())):
                raise ValueError(f"zip member escapes destination: {info.filename}")
            target.parent.mkdir(parents=True, exist_ok=True)
            with zf.open(info) as src, target.open("wb") as dst:
                shutil.copyfileobj(src, dst)
    return dest


def tail_text(path: Path, max_lines: int) -> str:
    try:
        lines = read_text_auto(path).splitlines()
    except (OSError, UnicodeDecodeError) as exc:
        return f"<failed to read {path}: {exc}>"
    return "\n".join(lines[-max_lines:])


def load_preflight(path: Path) -> dict[str, Any]:
    try:
        data = json.loads(read_text_auto(path))
    except (OSError, UnicodeDecodeError, json.JSONDecodeError) as exc:
        return {"path": str(path), "parsed": False, "error": str(exc)}
    if not isinstance(data, dict):
        return {"path": str(path), "parsed": False, "error": "preflight JSON was not an object"}
    checks = data.get("checks") if isinstance(data.get("checks"), list) else []
    failed_checks = [
        check.get("name")
        for check in checks
        if isinstance(check, dict) and check.get("ok") is False
    ]
    nvidia_smi = next(
        (check for check in checks if isinstance(check, dict) and check.get("name") == "nvidia_smi"),
        None,
    )
    return {
        "path": str(path),
        "parsed": True,
        "ok": data.get("ok"),
        "profile": data.get("profile"),
        "base_url": data.get("base_url"),
        "llama_server_found": data.get("llama_server_found"),
        "failures": data.get("failures") if isinstance(data.get("failures"), list) else [],
        "failed_checks": failed_checks,
        "nvidia_smi": nvidia_smi,
    }


def summarize_dir(report_dir: Path, log_tail_lines: int = 60) -> dict[str, Any]:
    preflight_paths = sorted(report_dir.glob("**/preflight-*.json"), key=lambda p: p.stat().st_mtime)
    server_logs = sorted(report_dir.glob("**/server-*.log"), key=lambda p: p.stat().st_mtime)
    preflights = [load_preflight(path) for path in preflight_paths]
    latest_preflight = preflights[-1] if preflights else None
    latest_log = server_logs[-1] if server_logs else None
    summary: dict[str, Any] = {
        "schema": SCHEMA,
        "report_dir": str(report_dir),
        "preflight_count": len(preflight_paths),
        "server_log_count": len(server_logs),
        "latest_preflight": latest_preflight,
        "latest_server_log": str(latest_log) if latest_log else "",
        "latest_server_log_tail": tail_text(latest_log, log_tail_lines) if latest_log else "",
    }
    if latest_preflight is None:
        summary["status"] = "NO_PREFLIGHT"
    elif latest_preflight.get("ok") is True:
        summary["status"] = "PREFLIGHT_OK"
    else:
        summary["status"] = "PREFLIGHT_FAILED"
    return summary


def import_report_bundle(args: argparse.Namespace) -> dict[str, Any]:
    receive = None if args.skip_taildrop else receive_taildrop(args.inbox, args.wait)
    archive = args.archive
    if archive is None:
        candidates = archive_candidates(args.inbox)
        if not candidates:
            return {
                "schema": SCHEMA,
                "imported": False,
                "receive": receive,
                "error": f"no qwen36-node-reports-*.zip bundle found under {args.inbox}",
            }
        archive = candidates[0]
    dest = extract_archive(archive, args.out_dir, args.replace)
    summary = summarize_dir(dest, log_tail_lines=args.log_tail_lines)
    summary.update({"imported": True, "archive": str(archive), "receive": receive})
    return summary


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--inbox", type=Path, default=DEFAULT_INBOX)
    parser.add_argument("--out-dir", type=Path, default=DEFAULT_OUT_DIR)
    parser.add_argument("--archive", type=Path, help="specific qwen36-node-reports-*.zip to import")
    parser.add_argument("--wait", action="store_true", help="wait for one Taildrop file before importing")
    parser.add_argument("--skip-taildrop", action="store_true", help="do not run tailscale file get first")
    parser.add_argument("--replace", action="store_true", help="replace an existing extracted report directory")
    parser.add_argument("--log-tail-lines", type=int, default=60)
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    try:
        summary = import_report_bundle(args)
    except Exception as exc:
        print(json.dumps({"schema": SCHEMA, "imported": False, "error": str(exc)}, indent=2))
        return 1
    print(json.dumps(summary, indent=2))
    return 0 if summary.get("imported") else 1


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: internal/qwen36nodereports/qwen36nodereports.go
package qwen36nodereports

import (
	"archive/zip"
	"bytes"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"unicode/utf16"
)

const Schema = "fak.qwen36-node-reports.v1"

type Receive struct {
	Ran      bool     `json:"ran"`
	Reason   string   `json:"reason,omitempty"`
	Command  []string `json:"command,omitempty"`
	ExitCode int      `json:"exit_code,omitempty"`
	Stdout   string   `json:"stdout,omitempty"`
	Stderr   string   `json:"stderr,omitempty"`
	Inbox    string   `json:"inbox"`
}

type ImportArgs struct {
	Inbox        string
	OutDir       string
	Archive      string
	Wait         bool
	SkipTaildrop bool
	Replace      bool
	LogTailLines int
}

func FindTailscale() string {
	for _, name := range []string{"tailscale", "tailscale.exe"} {
		if p, err := exec.LookPath(name); err == nil {
			return p
		}
	}
	candidate := `C:\Program Files\Tailscale\tailscale.exe`
	if _, err := os.Stat(candidate); err == nil {
		return candidate
	}
	return ""
}

func ReceiveTaildrop(inbox string, wait bool) Receive {
	_ = os.MkdirAll(inbox, 0o755)
	exe := FindTailscale()
	if exe == "" {
		return Receive{Ran: false, Reason: "tailscale CLI not found", Inbox: inbox}
	}
	cmd := []string{exe, "file", "get", "--conflict=rename"}
	if wait {
		cmd = append(cmd, "--wait")
	}
	cmd = append(cmd, inbox)
	proc := exec.Command(cmd[0], cmd[1:]...)
	out, err := proc.Output()
	stderr := ""
	code := 0
	if err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			stderr = string(ee.Stderr)
			code = ee.ExitCode()
		} else {
			stderr = err.Error()
			code = 127
		}
	}
	return Receive{Ran: true, Command: cmd, ExitCode: code, Stdout: tail(string(out), 2000), Stderr: tail(stderr, 2000), Inbox: inbox}
}

func ArchiveCandidates(inbox string) []string {
	var paths []string
	for _, pattern := range []string{"qwen36-node-reports-*.zip", "*qwen36-node-reports-*.zip"} {
		hits, _ := filepath.Glob(filepath.Join(inbox, pattern))
		paths = append(paths, hits...)
	}
	uniq := map[string]bool{}
	var out []string
	for _, p := range paths {
		if !uniq[p] {
			uniq[p] = true
			out = append(out, p)
		}
	}
	sort.Slice(out, func(i, j int) bool {
		ai, _ := os.Stat(out[i])
		aj, _ := os.Stat(out[j])
		if ai == nil || aj == nil {
			return out[i] < out[j]
		}
		return ai.ModTime().After(aj.ModTime())
	})
	return out
}

func SafeMemberName(name string) ([]string, error) {
	normalized := strings.ReplaceAll(name, "\\", "/")
	if strings.HasPrefix(normalized, "/") || strings.TrimSpace(normalized) == "" {
		return nil, fmt.Errorf("unsafe zip member path: %s", name)
	}
	parts := strings.Split(normalized, "/")
	var clean []string
	for _, part := range parts {
		if part == "" || part == "." {
			continue
		}
		if part == ".." || strings.Contains(part, ":") {
			return nil, fmt.Errorf("unsafe zip member path: %s", name)
		}
		clean = append(clean, part)
	}
	if len(clean) == 0 {
		return nil, fmt.Errorf("unsafe zip member path: %s", name)
	}
	return clean, nil
}

func ValidateZip(archive string) ([]*zip.File, error) {
	zr, err := zip.OpenReader(archive)
	if err != nil {
		return nil, err
	}
	defer zr.Close()
	return validateZipFiles(zr.File, archive)
}

func ExtractArchive(archive, outDir string, replace bool) (string, error) {
	abs, _ := filepath.Abs(archive)
	dest := filepath.Join(outDir, strings.TrimSuffix(filepath.Base(archive), filepath.Ext(archive)))
	if _, err := os.Stat(dest); err == nil && !replace {
		return "", fmt.Errorf("%s already exists; rerun with --replace", dest)
	}
	if err := os.RemoveAll(dest); err != nil {
		return "", err
	}
	if err := os.MkdirAll(dest, 0o755); err != nil {
		return "", err
	}
	zr, err := zip.OpenReader(abs)
	if err != nil {
		return "", err
	}
	defer zr.Close()
	members, err := validateZipFiles(zr.File, abs)
	if err != nil {
		return "", err
	}
	destAbs, _ := filepath.Abs(dest)
	for _, info := range members {
		parts, err := SafeMemberName(info.Name)
		if err != nil {
			return "", err
		}
		target := filepath.Join(append([]string{dest}, parts...)...)
		targetAbs, _ := filepath.Abs(target)
		if targetAbs != destAbs && !strings.HasPrefix(targetAbs, destAbs+string(os.PathSeparator)) {
			return "", fmt.Errorf("zip member escapes destination: %s", info.Name)
		}
		if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
			return "", err
		}
		src, err := info.Open()
		if err != nil {
			return "", err
		}
		dst, err := os.Create(target)
		if err != nil {
			_ = src.Close()
			return "", err
		}
		_, copyErr := io.Copy(dst, src)
		closeErr := errors.Join(src.Close(), dst.Close())
		if copyErr != nil {
			return "", copyErr
		}
		if closeErr != nil {
			return "", closeErr
		}
	}
	return dest, nil
}

func validateZipFiles(files []*zip.File, archive string) ([]*zip.File, error) {
	var members []*zip.File
	for _, f := range files {
		if _, err := SafeMemberName(f.Name); err != nil {
			return nil, err
		}
		if f.FileInfo().IsDir() {
			continue
		}
		members = append(members, f)
	}
	if len(members) == 0 {
		return nil, fmt.Errorf("archive contains no files: %s", archive)
	}
	return members, nil
}

func ReadTextAuto(path string) (string, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	if bytes.HasPrefix(raw, []byte{0xff, 0xfe}) {
		return decodeUTF16(raw[2:], binary.LittleEndian), nil
	}
	if bytes.HasPrefix(raw, []byte{0xfe, 0xff}) {
		return decodeUTF16(raw[2:], binary.BigEndian), nil
	}
	if bytes.HasPrefix(raw, []byte{0xef, 0xbb, 0xbf}) {
		raw = raw[3:]
	}
	if json.Valid(raw) || utf8ish(raw) {
		return string(raw), nil
	}
	return decodeUTF16(raw, binary.LittleEndian), nil
}

func TailText(path string, maxLines int) string {
	text, err := ReadTextAuto(path)
	if err != nil {
		return fmt.Sprintf("<failed to read %s: %v>", path, err)
	}
	lines := strings.Split(strings.ReplaceAll(text, "\r\n", "\n"), "\n")
	for len(lines) > 0 && lines[len(lines)-1] == "" {
		lines = lines[:len(lines)-1]
	}
	if maxLines > 0 && len(lines) > maxLines {
		lines = lines[len(lines)-maxLines:]
	}
	return strings.Join(lines, "\n")
}

func LoadPreflight(path string) map[string]any {
	text, err := ReadTextAuto(path)
	if err != nil {
		return map[string]any{"path": path, "parsed": false, "error": err.Error()}
	}
	var data map[string]any
	if err := json.Unmarshal([]byte(text), &data); err != nil {
		return map[string]any{"path": path, "parsed": false, "error": err.Error()}
	}
	checks, _ := data["checks"].([]any)
	var failed []string
	var nvidia any
	for _, item := range checks {
		check, ok := item.(map[string]any)
		if !ok {
			continue
		}
		if check["ok"] == false {
			if name, ok := check["name"].(string); ok {
				failed = append(failed, name)
			}
		}
		if check["name"] == "nvidia_smi" {
			nvidia = check
		}
	}
	failures, ok := data["failures"].([]any)
	if !ok {
		failures = []any{}
	}
	return map[string]any{
		"path":               path,
		"parsed":             true,
		"ok":                 data["ok"],
		"profile":            data["profile"],
		"base_url":           data["base_url"],
		"llama_server_found": data["llama_server_found"],
		"failures":           failures,
		"failed_checks":      failed,
		"nvidia_smi":         nvidia,
	}
}

func SummarizeDir(reportDir string, logTailLines int) map[string]any {
	preflightPaths, _ := filepath.Glob(filepath.Join(reportDir, "**", "preflight-*.json"))
	serverLogs, _ := filepath.Glob(filepath.Join(reportDir, "**", "server-*.log"))
	if len(preflightPaths) == 0 {
		preflightPaths = walkGlob(reportDir, "preflight-*.json")
	}
	if len(serverLogs) == 0 {
		serverLogs = walkGlob(reportDir, "server-*.log")
	}
	sortByMod(preflightPaths)
	sortByMod(serverLogs)
	preflights := make([]map[string]any, 0, len(preflightPaths))
	for _, p := range preflightPaths {
		preflights = append(preflights, LoadPreflight(p))
	}
	var latestPreflight any
	if len(preflights) > 0 {
		latestPreflight = preflights[len(preflights)-1]
	}
	latestLog := ""
	if len(serverLogs) > 0 {
		latestLog = serverLogs[len(serverLogs)-1]
	}
	summary := map[string]any{
		"schema":                 Schema,
		"report_dir":             reportDir,
		"preflight_count":        len(preflightPaths),
		"server_log_count":       len(serverLogs),
		"latest_preflight":       latestPreflight,
		"latest_server_log":      latestLog,
		"latest_server_log_tail": "",
	}
	if latestLog != "" {
		summary["latest_server_log_tail"] = TailText(latestLog, logTailLines)
	}
	if latestPreflight == nil {
		summary["status"] = "NO_PREFLIGHT"
	} else if lp, ok := latestPreflight.(map[string]any); ok && lp["ok"] == true {
		summary["status"] = "PREFLIGHT_OK"
	} else {
		summary["status"] = "PREFLIGHT_FAILED"
	}
	return summary
}

func ImportReportBundle(args ImportArgs) map[string]any {
	var receive any
	if !args.SkipTaildrop {
		receive = ReceiveTaildrop(args.Inbox, args.Wait)
	}
	archive := args.Archive
	if archive == "" {
		candidates := ArchiveCandidates(args.Inbox)
		if len(candidates) == 0 {
			return map[string]any{"schema": Schema, "imported": false, "receive": receive, "error": fmt.Sprintf("no qwen36-node-reports-*.zip bundle found under %s", args.Inbox)}
		}
		archive = candidates[0]
	}
	dest, err := ExtractArchive(archive, args.OutDir, args.Replace)
	if err != nil {
		return map[string]any{"schema": Schema, "imported": false, "receive": receive, "error": err.Error()}
	}
	summary := SummarizeDir(dest, args.LogTailLines)
	summary["imported"] = true
	summary["archive"] = archive
	summary["receive"] = receive
	return summary
}

func MarshalJSON(v any) ([]byte, error) {
	return json.MarshalIndent(v, "", "  ")
}

func tail(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[len(s)-n:]
}

func decodeUTF16(raw []byte, order binary.ByteOrder) string {
	u16 := make([]uint16, 0, len(raw)/2)
	for len(raw) >= 2 {
		u16 = append(u16, order.Uint16(raw[:2]))
		raw = raw[2:]
	}
	return string(utf16.Decode(u16))
}

func utf8ish(raw []byte) bool {
	for _, b := range raw {
		if b == 0 {
			return false
		}
	}
	return true
}

func walkGlob(root, pattern string) []string {
	var out []string
	_ = filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
		if err == nil && !d.IsDir() {
			if ok, _ := filepath.Match(pattern, filepath.Base(path)); ok {
				out = append(out, path)
			}
		}
		return nil
	})
	return out
}

func sortByMod(paths []string) {
	sort.Slice(paths, func(i, j int) bool {
		ai, _ := os.Stat(paths[i])
		aj, _ := os.Stat(paths[j])
		if ai == nil || aj == nil {
			return paths[i] < paths[j]
		}
		return ai.ModTime().Before(aj.ModTime())
	})
}
